contracts/integrations/nevermined/AgentMechSubscription.sol

```diff
@@ -0,0 +1,143 @@
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+/// @title AgentMechSubscription - Smart contract for extending AgentMech with subscription
+/// @dev A Mech that is operated by the holder of an ERC721 non-fungible token via a subscription.
+contract AgentMechSubscription is AgentMech {
+    event SubscriptionUpdated(address subscriptionNFT, uint256 subscriptionTokenId);
+
+    // Subscription NFT
+    address public subscriptionNFT;
+    // Subscription token Id
+    uint256 public subscriptionTokenId;
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    /// @dev AgentMechSubscription constructor.
+    /// @param _token Address of the token registry contract.
+    /// @param _tokenId The token ID.
+    /// @param _creditsPerRequest Number of credits to pay for request via subscription.
+    /// @param _subscriptionNFT Subscription address.
+    /// @param _subscriptionTokenId Subscription token Id.
+    constructor(
+        address _token,
+        uint256 _tokenId,
+        uint256 _creditsPerRequest,
+        address _subscriptionNFT,
+        uint256 _subscriptionTokenId
+    )
+        AgentMech(_token, _tokenId, _creditsPerRequest)
+    {
+        // Check for the subscription address
+        if (_subscriptionNFT == address(0)) {
+            revert ZeroSubscriptionAddress();
+        }
+
+        // Check for the subscription token Id
+        if (_subscriptionTokenId == 0) {
+            revert ZeroTokenId();
+        }
+
+        subscriptionNFT = _subscriptionNFT;
+        subscriptionTokenId = _subscriptionTokenId;
+    }
+
+    /// @dev Performs actions before the request is posted.
+    /// @param amount Amount of payment in wei.
+    function _preRequest(uint256 amount, uint256, bytes memory) internal override {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Check that there is no incoming deposit
+        if (amount > 0) {
+            revert NoDepositAllowed(amount);
+        }
+
+        // Check for the number of credits available in the subscription vs total number of credits needed
+        uint256 creditsBalance = IERC1155(subscriptionNFT).balanceOf(msg.sender, subscriptionTokenId);
+        uint256 numUndeliveredRequests = mapUndeliveredRequestsCounts[msg.sender];
+        uint256 creditsPerPendingRequests = (numUndeliveredRequests + 1) * price;
+        if (creditsBalance < creditsPerPendingRequests) {
+            revert NotEnoughCredits(creditsBalance, creditsPerPendingRequests);
+        }
+
+        _locked = 1;
+    }
+
+    /// @dev Performs actions before the delivery of a request.
+    /// @param requestId Request Id.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @return requestData Data for the request processing.
+    function _preDeliver(uint256 requestId, bytes memory data) internal override returns (bytes memory requestData) {
```

Comment:
should take a `cost` argument. Which should be used in the burn below. This will require a new deliver method.

Reply:
This will be extracted from the data itself